### DISCUSSION TITLE
Redirecting after login

### DISCUSSION THREAD

**Discussion by Cody-W-Tucker** [vikaspotluri123/discourse-on-ghost#67] [Ideas]
I love the work you've done it's very impressive. I wondered if people are supposed to be redirected to the discourse landing page once they've been logged in. I don't see anything in the docs or env that would allow this. Have I missed something, or is this something that should be handled on my end?

**Comment by vikaspotluri123** [comment]
Hey @Cody-W-Tucker, thanks for asking! Redirection should be pretty transparent to everyone. Here's the sign in flow for reference:

1. Discourse receives a login request
2. Internally, Discourse tracks where this login request came from
3. Since Discourse Connect is enabled, Discourse redirects to DoG
4. DoG receives the login request, and redirects the user back to Discourse. The URL will contain encrypted member information
5. Discourse receives the login request, decrypts the information, and logs the user in
6. Discourse redirects the user
   - Note: I haven't reviewed exactly how Discourse handles this logic, but one of my tests shows that the user is redirected to where they logged in from

To summarize: Discourse is ultimately responsible for redirecting the user after DoG authenticates the user.

Hope this helps 😊

**Reply by Cody-W-Tucker** [reply]
I must have setup incorrectly then. 

For me the login process is this.

If a user clicks login on discourse, it sends users to the ghost's portal login screen, then sends a magic link to complete login dropping the user back on the ghost homepage.

When visiting discourse again the login button still shows, but when clicked it logs in automatically.

**Reply by vikaspotluri123** [reply]
> If a user clicks login on discourse, it sends users to the ghost's portal login screen, then sends a magic link to complete login dropping the user back on the ghost homepage.

If you're not already logged into Ghost, the login process _can_ be a bit wonky. From my understanding, the magic-link should include the "next" URL via the `r` parameter, which should handle most cases.

- Which auth mechanism are you using?
- What is your Ghost version?
- How are your users prompted to log in (via the theme, or via portal)?
- Do you have any extensions that could be stripping the `r` param from the magic-link?

**Reply by Cody-W-Tucker** [reply]
Auth: Magic link
Ghost Version: 5.51.2
Users log in via portal

I am using click tracking via Mailgun which uses a subdomain to track clicks. But the magic link doesn't appear to route through the subdomain. 

It actually appears that it redirects to the account page. r=https%3A%2F%2Fmy-domain.com%2F%23%2Fportal%2Faccount

**Reply by vikaspotluri123** [reply]
My bad, by auth I meant the SSO Type 😄 Which [SSO Type](https://github.vikaspotluri.me/discourse-on-ghost/configuration/) are you using?

> It actually appears that it redirects to the account page. r=https%3A%2F%2Fmy-domain.com%2F%23%2Fportal%2Faccount

Thanks, this suggests the redirect is being lost between DoG and logging into Ghost 👍

**Reply by Cody-W-Tucker** [reply]
Session, what do you think is causing the redirect loss?

**Reply by vikaspotluri123** [reply]
I think it's a bug with session auth; when redirecting to the login page, query params aren't included, I can create a beta release later today for you to try

**Reply by Cody-W-Tucker** [reply]
That would be awesome, happy to try it!

**Reply by vikaspotluri123** [reply]
Alright, here it is: https://github.com/vikaspotluri123/discourse-on-ghost/releases/tag/v0.2.2-beta.0

You can test it by running `npm install @potluri/discourse-on-ghost@0.2.2-beta.0`

Note: for a fully seamless experience, you might want to create a custom landing page (this applies to both session and jwt SSO), like [this example](https://github.com/vikaspotluri123/discourse-on-ghost/blob/v0.2.2-beta.0/landing-pages/not-logged-in.hbs). Otherwise, the user will be logged into Ghost, but won't be routed back to DoG to complete the Discourse login process.

**Comment by Cody-W-Tucker** [comment]
I installed the patch and restarted the service but the result was the same as before. But the `r` param was different.

Before: [r=https%3A%2F%2Fdomain.com%2F%23%2Fportal%2Faccount](https://codywtucker.com/members/?token=eZCofOiTAJi6GYdfgH43XmAsGfDkxGZH&action=signin&r=https%3A%2F%2Fdomain.com%2F%23%2Fportal%2Faccount)
After: [r=https%3A%2F%2Fdomain.com%2F%3Fsso%3Dbm9uY2U9ZTBhNTdhYjAxYjIwNWY0Mjc5OTIxZGVmMDU0N2UyZGYmcmV0dXJuX3Nzb191cmw9aHR0cHMlM0ElMkYlMkZkaXNjb3Vyc2UuY29keXd0dWNrZXIuY29tJTJGc2Vzc2lvbiUyRnNzb19sb2dpbg%253D%253D%26sig%3D4f4cd18328dcbbfe4d9fceb3146ea43a0b14cbb6dbb6679ca3f4d702af5992bb%23%2Fportal%2Faccount](https://domain.com/members/?token=NOUo-ZInTX5-l3u63xnOuMXSN6koeSpA&action=signin&r=https%3A%2F%2Fdomain.com%2F%3Fsso%3Dbm9uY2U9ZTBhNTdhYjAxYjIwNWY0Mjc5OTIxZGVmMDU0N2UyZGYmcmV0dXJuX3Nzb191cmw9aHR0cHMlM0ElMkYlMkZkaXNjb3Vyc2UuY29keXd0dWNrZXIuY29tJTJGc2Vzc2lvbiUyRnNzb19sb2dpbg%253D%253D%26sig%3D4f4cd18328dcbbfe4d9fceb3146ea43a0b14cbb6dbb6679ca3f4d702af5992bb%23%2Fportal%2Faccount)

**Reply by vikaspotluri123** [reply]
Perfect! Is it launching the Portal when the page loads? If not, you'll need to add at least [this portion of the script](https://github.com/vikaspotluri123/discourse-on-ghost/blob/ca6681cfd917c28f4b95b84885e3dd44fc0a41d0/landing-pages/not-logged-in.hbs#L84-L101) to your code injection.

**Reply by Cody-W-Tucker** [reply]
It's launching the portal when the page loads, but it's still not redirecting to discourse. After login, it shows logged on in ghost via portal/account.

**Reply by vikaspotluri123** [reply]
Do you have any browser errors in your console?

You might want to create a dedicated sign in page (e.g. `/sso`) with the `not-logged-in` template, and set `DOG_SSO_NO_AUTH_REDIRECT=https://your.domain/sso/` in your config to prevent Portal from popping up. Or you can add a script to code injection that strips the portal hash from the URL after portal opens.

**Reply by Cody-W-Tucker** [reply]
No browser errors, but when I create the page and set that option the discourse redirect login button sends me to the /sso/ page automatically. I'm guessing that means the auth isn't working.

Here's my .env because I think I've done something wrong.

DOG_HOSTNAME="127.0.0.1" # Optional
DOG_PORT="3286" # Optional
DOG_GHOST_URL="https://domain.com"
DOG_GHOST_ADMIN_URL="" # Optional - defaults to DOG_GHOST_URL
DOG_DISCOURSE_SHARED_SECRET="hidden"
DOG_GHOST_ADMIN_TOKEN="hidden"
DOG_DISCOURSE_URL="https://discourse.domain.com"
DOG_DISCOURSE_API_KEY="hidden"
DOG_DISCOURSE_API_USER="" # Optional
DOG_LOG_DISCOURSE_REQUESTS="false"
DOG_LOG_GHOST_REQUESTS="false"
DOG_GHOST_MEMBER_WEBHOOKS_ENABLED="false"
DOG_GHOST_MEMBER_UPDATED_WEBHOOK_ID="hidden"
DOG_GHOST_MEMBER_DELETED_WEBHOOK_ID="hidden"
DOG_GHOST_MEMBER_DELETE_DISCOURSE_ACTION="suspend"
DOG_DISCOURSE_SSO_TYPE="session"
DOG_JWT_GHOST_SSO_PAGE="/sso/" # Optional (absolute path)
DOG_SSO_NO_AUTH_REDIRECT="https://domain.com/sso/" # Optional (url)

**Reply by vikaspotluri123** [reply]
Just to be safe, can you make `DOG_SSO_NO_AUTH_REDIRECT="https://domain.com/sso/"` - I think Ghost strips query params if you don't include the trailing slash.

The expected behavior in this scenario:
1. Click login on Discourse
2. Get redirected to DoG. Not logged in so...
3. Get redirected to SSO page. SSO page has a log in button that launches portal
4. Enter email address to get magic-link
5. Click magic-link, get redirected back to SSO page. Since you're logged in, the javascript should catch that and redirect you back to DoG
6. DoG sees you're logged in and redirects you to Discourse

**Comment by Cody-W-Tucker** [comment]
I got this working today, it was on my end. I didn't understand that I needed to make a new page using the not-logged-in template. Thanks for walking me through this.

**Comment by mhpaler** [comment]
@vikaspotluri123 Not meaning to hijack this conversation, but I feel like given the topic of redirections, this is the right place.

The above discussion centers on the workflow where a user click a login link on Discourse and is redirected to DoG which then redirects to the ghost custom /sso to acquire a magic link.

I'm curious if you have thoughts one more use case on login workflow:

"As a user that is completely logged out of both Ghost and Discourse, if I login from directly from Ghost and I navigate from Ghost to Discourse, I would like to be automatically logged into Discourse." 

Currently, once logged into Ghost, navigating to Discourse requires clicking the login link on Discourse to login.

**Reply by mhpaler** [reply]
I actually have a configuration only solution to this problem, which I'll post later today.

**Reply by vikaspotluri123** [reply]
Hey @mpaler 👋🏾, this is a bit of a tricky ask, specifically because the UX would be a bit confusing:

 - Let's say a user logs in to Ghost
 - Since Portal handles user auth, let's assume it's possible for code injection to identify that a user just logged in
 - Now that we know a user logged in, we can run the Discourse login flow (Discourse Login &rarr; DoG &rarr; Ghost &rarr; Discourse &rarr; Logged in)
 - However, I don't think it's possible to redirect the user back to Ghost without some kind of hack.

One potential option is to disable anonymous access (`login required`) in the Discourse settings - which shifts forum auth to the first time the user accesses the forum, and should be pretty transparent.
